### ISSUE TITLE
Inconsistent escaping of Macro output.

### ISSUE THREAD

**Issue by legutierr** [pallets/jinja#565]
Macro output is supposed to be safe, but it is not in some circumstances.  For instance:

```
>>> env = jinja2.Environment(autoescape=False, extensions=[jinja2.ext.autoescape])                 
>>> template = "{% macro m() %}<html>{% endmacro %}"
>>> template += "{% autoescape true %}{{ m() }}{% endautoescape %}"
>>> env.from_string(template).render()
u'&lt;html&gt;'
```

Also:

```
>>> env = jinja2.Environment(autoescape=False)                 
>>> template = "{% macro m() %}<html>{% endmacro %}{{ m()|e }}"
>>> env.from_string(template).render()
u'&lt;html&gt;'
```

Only by changing the global autoescape value to true, is the macro output now safe:

```
>>> env = jinja2.Environment(autoescape=True, extensions=[jinja2.ext.autoescape])
>>> template = "{% macro m() %}<html>{% endmacro %}"
>>> template += "{% autoescape true %}{{ m() }}{% endautoescape %}"
>>> env.from_string(template).render()
u'<html>'
```

And:

```
>>> env = jinja2.Environment(autoescape=True)                 
>>> template = "{% macro m() %}<html>{% endmacro %}{{ m()|e }}"
>>> env.from_string(template).render()
u'<html>'
```

Interestingly, this problem doesn't exist if the macro definition occurs inside the autoescape block:

```
>>> env = jinja2.Environment(autoescape=False, extensions=[jinja2.ext.autoescape])
>>> template = "{% autoescape true %}{% macro m() %}<html>{% endmacro %}"
>>> template += "{{ m() }}{% endautoescape %}"
>>> env.from_string(template).render()
u'<html>'
```

**Comment by mitsuhiko** [comment]
This might be a problem with the optimization code in the compiler.

**Comment by mitsuhiko** [comment]
This actually is intentional now that I look at it. The autoescaping code is disabled if set to `False`. You can however set it to a function that returns False: `autoescape=lambda x: False` and the initial value of the autoescape flag is set accordingly.

**Comment by legutierr** [comment]
Hi @mitsuhiko.  

Could you provide more explanation of your decision to close this ticket?  I'm finding it difficult to understand why this is desirable behavior with regards to the macro escaping.  Shouldn't macro output _always_ be safe?  

And even if you want macro output to only _sometimes_ be safe, isn't the following behavior the reverse of what you would want in that case?

    >>> env = jinja2.Environment(autoescape=False, extensions=[jinja2.ext.autoescape])                 
    >>> template = "{% macro m() %}<html>{% endmacro %}"
    >>> template += "{% autoescape true %}{{ m() }}{% endautoescape %} {{ m()|e }}"
    >>> env.from_string(template).render()
    u'&lt;html&gt; &lt;html&gt;'

If the autoescape tag somehow escapes macro output by default, why would turning autoescape _on_ globally cause macros _not_ to be escaped?

    >>> env = jinja2.Environment(autoescape=True, extensions=[jinja2.ext.autoescape])                 
    >>> template = "{% macro m() %}<html>{% endmacro %}"
    >>> template += "{% autoescape true %}{{ m() }}{% endautoescape %} {{ m()|e }}"
    >>> env.from_string(template).render()
    u'<html> <html>'

Could it not be that there is some bug triggered by global autoescape=False that somehow is marking macros as unsafe when they should simply be safe in all cases?

**Comment by mitsuhiko** [comment]
This has actually nothing to do with macros. This is an unfortunate effect of how auto escaping is implemented.

The `autoescape` flag does not *just* enable the default setting of the value but also enables the feature in the first place.  The `jinja2.ext.autoescape` extension which changes the value of the autoescape flag requires `autoescape` to be enabled.

`autoescape` can be enabled in two ways: It can be set to `true` which enables the autoescape feature and sets it to "perform autoescaping". It can be set to `false` which disables auto escape mode and disables autoescaping entirely. Unfortunately this does not block the use of the `jinja2.ext.autoescape` extension which will perform incorrectly in this case.

The third mode is to set `autoescape` to a callable. Then autoescaping is enabled as a feature and the initial value of autoescaping for a template is decided by the return value of the function. So the correct behavior to enable the autoescape feature but disable it by default for a template is to set `autoescape=lambda x: False`.

I hope that explains the feature better.

**Comment by mitsuhiko** [comment]
This btw has nothing to do with macros. This would come up in other situations as well.

**Comment by mitsuhiko** [comment]
I can see that behavior being confusing however. I will investigate if I can improve on it so that the compiler support for autoescape does not depend on the initial flag.

**Comment by mitsuhiko** [comment]
Going to reopen this and investigate if we can improve here.

**Comment by legutierr** [comment]
Hi @mitsuhiko, thank you for the prompt response, and thank you for re-opening.

With regards to the feature you describe, I'm not sure if I see it in this particular example.

    >>> env = jinja2.Environment(autoescape=False, extensions=[jinja2.ext.autoescape])                 
    >>> template = "{% macro m() %}<html>{% endmacro %}"
    >>> template += "{% autoescape true %}{{ m() }}{% endautoescape %} {{ m()|e }} {{ m() }}"
    >>> env.from_string(template).render()
    u'&lt;html&gt; &lt;html&gt; <html>'

In this example, global autoescape is turned off, but the autoescape tag is doing something anyway.  It doesn't seem to be disabled by the global autoescape being False (which, to me, seems like a good thing, except with regards to this macro issue).  Also, the escape filter is affected by the global autoescape setting, so the underlying issue doesn't seem to be limited to the autoescape extension.

**Comment by mitsuhiko** [comment]
Turns out this is actually not even the case I had in mind and you are right that this makes no sense. I fixed that now. This actually turns out to be an issue with the macro deciding at compile time what the autoescape flag will be which is changed at the time the call happens.

The code already dealt with autoescape being flipped at a later point through the volatile flag which seems to work from what I can tell.
